crates/sparrow-instructions/src/evaluators/list/collect_struct.rs

```diff
@@ -395,16 +419,142 @@ impl CollectStructEvaluator {
+        assert_eq!(entity_indices.len(), input_structs.len());
+
+        Self::ensure_entity_capacity(token, key_capacity)?;
+
+        // Recreate the take indices for the current state
+        let mut entity_take_indices = Self::construct_entity_take_indices(token);
+
+        let old_state = token.state.as_list::<i32>();
+        let old_state_flat = old_state.values();
+
+        let mut take_output_builder = UInt32Builder::new();
+        let mut output_offset_builder = vec![0];
+
+        // Tracks the result's null values
+        let mut null_buffer = vec![];
+
+        // Concat the state's times and the input's times
+        let old_times = token.times.as_list::<i32>();
+        let old_times_flat = old_times.values();
+        let combined_times =
+            arrow::compute::concat(&[old_times_flat.as_ref(), input_times.as_ref()])?;
+        let combined_times: &TimestampNanosecondArray = combined_times.as_primitive();
+        assert_eq!(
+            old_state_flat.len(),
+            old_times_flat.len(),
+            "time and state length mismatch"
+        );
+
+        let mut cur_offset = 0;
+        // For each entity, append the take indices for the new input to the existing
+        // entity take indices
+        for (index, entity_index) in entity_indices.values().iter().enumerate() {
+            // Update state
+            if input.is_valid(index) {
+                let take_index = (old_state_flat.len() + index) as u32;
+                entity_take_indices
+                    .entry(*entity_index)
+                    .and_modify(|v| {
+                        v.push_back(take_index);
+                        if v.len() > max {
+                            v.pop_front();
+                        }
+
+                        // Iterates over the front of the vec and pops off any values that
+                        // exist prior to the start of the current window
+                        if let Some(front_i) = v.front() {
+                            let mut oldest_time = combined_times.value(*front_i as usize);
+                            // Note this uses the `combined_times` and `take_index`
+                            // because it's possible we need to pop off new input
+                            let min_window_start =
+                                combined_times.value(take_index as usize) - duration;
+                            while oldest_time < min_window_start {
+                                v.pop_front();
+                                // Safety: we know there's elements in the vec, and we can't
+                                // have popped the last element because we just added one at a time
+                                // greater than the `min_window_start`.
+                                let front_i = v.front().unwrap();
+                                oldest_time = combined_times.value(*front_i as usize);
+                            }
+                        }
```

Comment:
There might be a better solution where I map over the `take_output_builder` and `take_output_indices`, and do the slicing to time there, but this works for now (I think!)